test/controllers/users_controller_test.rb

```diff
@@ -0,0 +1,74 @@
+require "test_helper"
+
+describe UsersController do
+  before do
+    @user = User.first
+    @login_data = {
+      user: {
+        username: @user.username,
+      },
+    }
+  end
+
+  describe "index" do
+    it "can get index" do
+      get users_path
+
+      must_respond_with :success
+    end
+  end
+
+  describe "show" do
+    it "returns a 404 status code if the user doesn't exist" do
+      user_id = 12345392487
+      expect(User.find_by(id: user_id)).must_be_nil
+
+      get user_path(user_id)
+
+      must_respond_with :not_found
+    end
+
+    it "functions for a user that exists" do
+      get user_path(@user.id)
+
+      must_respond_with :success
+    end
+  end
+
+  describe "login_form" do
+    it "can get the login page" do
+      get login_path
+
+      must_respond_with :success
+    end
+  end
+
+  describe "login" do
+    it "will login user" do
+      post login_path, params: @login_data
```

Comment:
You have two tests here, but these are only covering one behavior mode for this controller action (logging in an existing user). I would probably combine these into one.

However, there are a couple of behavior modes you're not covering:
- Logging in a new user (should create that user)
- Attempting to log in with an invalid username (e.g. empty string) (should fail)